n_queen_problem/n_queen_problem.py

```diff
@@ -0,0 +1,60 @@
+"""
+Following code is the implementation of the N-queen problem.
+The solution is to place the queens on board such that
+no queen is under attack by another queen.
+This implementation computes all solutions for a board,
+then prints the number of solutions. The number of solutions
+follows the OEIS sequence https://oeis.org/A000170.
+"""
+
+
+def under_attack(col, queens):
+    """
+    :param col: int
+        Column of queen to test
+    :param queens: list of tuple of ints
+        List of coordinates of other queens
+    :return: bool
+        True iff queen in given position is under attack
+    """
+
+    left = right = col
+    for r, c in reversed(queens):
+        left, right = left - 1, right + 1
+        if c in (left, col, right):
+            return True
+    return False
+
+
+def solve(board_size, queens):
+    """
+    :param board_size: int
+        Size of board to solve
+    :param queens: int
+        Number of queens to place in board
+    :return: list of list of tuple of ints
+        Each list contains the coordinates of the queens
+        that solve the problem
+    """
+
+    if queens == 0:  # trivial case
+        return [[]]
+
+    smaller_solutions = solve(board_size, queens - 1)
+    solutions = []
+    for solution in smaller_solutions:
+        for column in range(1, board_size + 1):
+            if not under_attack(column, solution):
+                solutions.append(solution + [(queens, column)])
+
+    return solutions
+
+
+def main():
+    board_size = 10
+    solutions = solve(board_size, board_size)
```

Comment:
This implementation is more general .. given a 8 x 8 board you could want to find the solutions with 7 queens .. should I revert to the original version?